File: Fortran/Types/Reduce/fails.f90
PROGRAM type

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100
  INTEGER, PARAMETER :: nitems = 100

  INTEGER :: rank
  INTEGER :: nproc
  INTEGER :: ierr
  INTEGER :: contig_type, custom_sum
  INTEGER, DIMENSION(nitems) :: values, values_recv, values_recv_ref

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  values = rank

  !Create the type
  CALL MPI_Type_contiguous(nitems, MPI_INTEGER, contig_type, ierr)

  !Register the type
  CALL MPI_Type_commit(contig_type, ierr)

  CALL MPI_Reduce(values, values_recv, 1, contig_type, MPI_SUM, 0, &
      MPI_COMM_WORLD, ierr)

  CALL MPI_Reduce(values, values_recv_ref, nitems, MPI_INTEGER, MPI_SUM, 0, &
      MPI_COMM_WORLD, ierr)

  IF (rank == 0) THEN
    PRINT *,"Sum over all processors = ", SUM(values_recv)
    PRINT *,"Sum over all processors (reference version) = ", &
        SUM(values_recv_ref)
  END IF

  CALL MPI_Finalize(ierr)

END PROGRAM type

File: C/Types/Reduce/fails.c
#include <stdio.h>
#include <mpi.h>

#define NITEMS 100

int main(int argc, char** argv)
{

  int rank, recv_rank, nproc, left, right, iitems, sum, sum_recv;
  int values[NITEMS], values_recv[NITEMS], values_recv_ref[NITEMS];
  MPI_Datatype contig_type;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  for (iitems = 0; iitems < NITEMS; ++iitems){
    values[iitems] = rank;
  }

  MPI_Type_contiguous(NITEMS, MPI_INT, &contig_type);
  MPI_Type_commit(&contig_type);

  if (rank == 0) printf("MPI_Type_contiguous used as send and recieve types\n");

  MPI_Reduce(values, values_recv, 1, contig_type, MPI_SUM, 0, MPI_COMM_WORLD);
  MPI_Reduce(values, values_recv_ref, NITEMS, MPI_INT, MPI_SUM, 0,
      MPI_COMM_WORLD);

  MPI_Type_free(&contig_type);

  sum = 0;
  sum_recv = 0;
  for (iitems = 0; iitems < NITEMS; ++iitems){
    sum += values_recv[iitems];
    sum_recv += values_recv[iitems];
  }

  if (rank == 0) {
    printf("Sum over all processors = %5d", sum);
    printf("Sum over all processors (reference version) = %5d", sum_recv);
  }

  MPI_Finalize();

}
